Fixes for MTK v8

ValentinKaisermayer opened the pull request:
# Changes
-  `Blocks` make use of a `connector` in the form of a `RealInput` and `RealOutput` (as Modelica does). However, this could be improved since there is only one variable in the connector. Definitions of `input=true` and `output=true` where removed since this feature is not fully developed in MTK. Can be added later in one place.
- Where applicable `Electrical` components make use of `OnePort`.
- Where applicable `Thermal` components make use of `Element1D`.

# Still open
- [ ] Tests for `Magnetic`
- [x] `PID`-controller and `StateSpace` block
- [x] A bunch of math blocks e.g. `Tan`
- [ ] A bunch of source blocks e.g. `Square`
- [x] Doc strings
- [x] Port names, parameter names and component names should be as in the Modelica standard library. The goal would be that existing Modelica examples can be almost copy pasted.

ChrisRackauckas commented:
Can we get to a merge point and then do a different PR to update the other things? I think too many PRs in this repo have been big "fix all of the things" PRs that never got merged, that then stopped other people from contributing as they waited for someone to finish a "fix all of the things" PR. So a "here's MTK v8" with 7 issues of harder things to update, that's a much better option in my book IMO, unless you plan to really knock out that list in a week.

ValentinKaisermayer commented:
I got a bit carried away. 😄 

I can give you a PR with just the connector stuff from v8 but I think that will not be enough for the tests to pass.

At this point it might be easier to fix the rest of the stuff.

ChrisRackauckas commented:
Mark some  tests as broken if you need to. Anything to restart the development here is appreciated.

ChrisRackauckas commented on src/Blocks/continuous.jl:
```diff
@@ -214,28 +240,26 @@ y = Cx + Du
         y .~ C*x .+ D*u
     ]
-    ODESystem(eqs, t, [x,y,u], [], name=name)
+    extend(ODESystem(eqs, t, vcat(x...), [], name=name), mimo)
```
`reduce(vcat,x)`

ValentinKaisermayer commented:
I will add some tests for the magnetic domain but will leave the digital for a later PR. It was commented out in v0.1.0 anyway.

ValentinKaisermayer commented on src/Blocks/utils.jl:
```diff
@@ -0,0 +1,47 @@
+@connector function RealInput(;name, nin=1, u_start=nin > 1 ? 0.0 : zeros(nin))
```
I added `RealInput` and `RealOutput` as connectors, such that the components can be connected via `connect`. However, the connectors only have one variable, which makes accessing them a bit redundant. Maybe a this can be handled through MTK via a `connect` for `Num` and just using a plain variable. For the blocks there is no need for a sophisticated stuff like inside and outside connectors or balance equations.

baggepinnen reviewed (commented):
It looks like a lot of the tests that tested that the solution of a simulation was actually correct have been removed. What's the motivation for removing those tests?

ValentinKaisermayer commented on src/Blocks/continuous.jl:
```diff
@@ -43,14 +30,16 @@ and a state-space realization is given by `ss(-1/T, 1/T, -k/T, k/T)`
-function Derivative(; k=1, T, name)
-    @variables x(t)=0 u(t)=0 [input=true] y(t)=0 [output=true]
-    @parameters T=T k=k
+function Derivative(; name, k=1, T=10, x0=0)
```
The Modelica default is `0.01`.

ValentinKaisermayer replied:
I removed the default value

ValentinKaisermayer commented on src/Blocks/continuous.jl:
```diff
@@ -142,28 +194,31 @@ function PID(; k, Ti=false, Td=false, wp=1, wd=1,
-        I.u ~ e + 1/(k*Ni)*ea  # Connect integrator block. The integrator integrates the control error and the anti-wind up tracking. Note the apparent tracking time constant 1/(k*Ni), since k appears after the integration and 1/Ti appears in the integrator block, the final tracking gain will be 1/(Ti*Ni) 
-        sat.u ~ derivative_action ? k*(ep + I.y + D.y) : k*(ep + I.y) # unsaturated output = P + I + D
+        e ~ r.u - y.u # Control error
+        ep ~ wp * r.u - y.u  # Control error for proportional part with setpoint weight
```
That is the problem I was referring to earlier. The connector only has one variable `u`, so accessing the variable means accessing `u` but you have to explicitly do it. In Modelica there is the possibility to define [a connector without any variables](https://github.com/modelica/ModelicaStandardLibrary/blob/master/Modelica/Blocks/Interfaces.mo#L7) and use it in place of a variable. In contrast to [connectors with variables](https://github.com/modelica/ModelicaStandardLibrary/blob/master/Modelica/Electrical/Analog/Interfaces/Pin.mo#L2). MTK does not have this functionality right now.

ValentinKaisermayer commented:
> It looks like a lot of the tests that tested that the solution of a simulation was actually correct have been removed. What's the motivation for removing those tests?

I have added checks for the solutions. 

My goal was that the tests additionally function as examples and in most cases the first part of each `@testset` may be simply copied by any user.

ValentinKaisermayer commented on src/Blocks/math.jl:
```diff
@@ -1,49 +1,305 @@
+"""
+Output the square root of the input (input >= 0 required).
+"""
+function Sqrt(;name)
```
You are right but I think this is a design decision. Do we want to stay close to the Modelica standard library or not. I think we should. For example Simulink has all of those blocks and a custom function block as well.

ValentinKaisermayer replied:
Than let's do that. BTW, we have the same problem with voltage and current sources, the base implementation already exists in Blocks.

ChrisRackauckas replied:
I think it's fine to have this if someone has already written it. It makes the translation a little easier / more direct. I wouldn't've spent the time to actually write the block, but if it exists then we might as well have a `ModelingToolkitStandardLibrary.Math` module that has all of these basic things, and they wouldn't bloat the namespace unless someone `using`s it.

ValentinKaisermayer commented on src/Blocks/math.jl:
```diff
@@ -1,49 +1,305 @@
-    Sum(k::AbstractVector; name)
+Output the sum of the three inputs.
+"""
+function Add3(;name, k1=1, k2=1, k3=1)
```
Sum still exists. These blocks have 2 or 3 scalar inputs and not a vector as input. That is not the same.

jaakkor2 commented on src/Thermal/HeatTransfer/sensors.jl:
```diff
@@ -1,36 +1,57 @@
+"""
+Absolute temperature sensor in Kelvin.
```
Names of units should be written in lowercase, e..g, kelvin.  See for example https://cdn.journals.aps.org/files/styleguide-pr.pdf sectionj III B.

ChrisRackauckas commented:
Deleting the tests is definitely a bad thing. We won't be able to merge this with those older tests removed.

ValentinKaisermayer commented:
> Deleting the tests is definitely a bad thing. We won't be able to merge this with those older tests removed.

I re-added explicit tests for the sim results of the analog sources, first and second order systems. I also added them for the block sources. I hope this will be sufficient. 

I also added a first test for the magnetic domain. However, most of the magnetic components are still not covered.

ChrisRackauckas commented:
The only tests I see are missing are the digital tests. Those are commented out. With those back I think this is ready to merge.

ValentinKaisermayer commented:
> The only tests I see are missing are the digital tests. Those are commented out. With those back I think this is ready to merge.

They are commented out in the current release as well.

ChrisRackauckas commented:
Okay cool, amazing. Just noticed that. I think it's fine to merge this and call it v1.0, and further improvements will continue to happen, but this is a major step forward already and at least 1000x better than what we had before. Major props to @ValentinKaisermayer !
